Subject: [PATCH] Fix compilation when osx <= 10.7

* AppKitImpl.mm: Syntax is not understood by 10.7, update it to be understandable by <= 10.7
error: expected method to read dictionary element not found on object of type 'NSDictionary *'
    NSRunningApplication* app = userInfo[NSWorkspaceApplicationKey];
                                ^
--- src/gui/osutils/macutils/AppKitImpl.mm
+++ src/gui/osutils/macutils/AppKitImpl.mm
@@ -60,7 +60,7 @@ - (id) initWithObject:(AppKit*)appkit
 - (void) didDeactivateApplicationObserver:(NSNotification*) notification
 {
     NSDictionary* userInfo = notification.userInfo;
-    NSRunningApplication* app = userInfo[NSWorkspaceApplicationKey];
+    NSRunningApplication* app = [userInfo objectForKey:NSWorkspaceApplicationKey];
 
     if (app.processIdentifier != [self ownProcessId]) {
         self.lastActiveApplication = app;
--
Subject: [PATCH] Fix compilation when osx <= 10.9

* AppKitImpl.mm: The code uses @available syntax which is new in clang 9
This prevents compilation on older version of macOS that don't
use this version. For example on El Capitan.

* AppKitImpl.mm: button property is new in 10.10. It is used for a feature of KeePassXC
  that is only available from 10.17 onwards. So we don't need it when compiling on <= 10.9
error: property 'button' not found on object of type 'NSStatusItem *'
        NSString* appearance = [dummy.button.effectiveAppearance.name lowercaseString];
                                      ^
--- src/gui/osutils/macutils/AppKitImpl.mm
+++ src/gui/osutils/macutils/AppKitImpl.mm
@@ -139,6 +139,7 @@ - (bool) isDarkMode
 //
 - (bool) isStatusBarDark
 {
+#if __clang_major__ >= 9 && MAC_OS_X_VERSION_MIN_REQUIRED >= 101000
     if (@available(macOS 10.17, *)) {
         // This is an ugly hack, but I couldn't find a way to access QTrayIcon's NSStatusItem.
         NSStatusItem* dummy = [[NSStatusBar systemStatusBar] statusItemWithLength:0];
--- src/gui/osutils/macutils/AppKitImpl.mm
+++ src/gui/osutils/macutils/AppKitImpl.mm
@@ -146,6 +147,7 @@ - (bool) isStatusBarDark
         [[NSStatusBar systemStatusBar] removeStatusItem:dummy];
         return [appearance containsString:@"dark"];
     }
+#endif
 
     return [self isDarkMode];
 }
--
Subject: [PATCH] Fix compilation when osx <= 10.8

* AppKitImpl.mm: AXIsProcessTrustedWithOptions exists from 10.9 onwards
error: use of undeclared identifier 'kAXTrustedCheckOptionPrompt'
error: use of undeclared identifier 'AXIsProcessTrustedWithOptions'
--- src/gui/osutils/macutils/AppKitImpl.mm
+++ src/gui/osutils/macutils/AppKitImpl.mm
@@ -168,9 +168,15 @@ - (void) userSwitchHandler:(NSNotification*) notification
 //
 - (bool) enableAccessibility
 {
-    // Request accessibility permissions for Auto-Type type on behalf of the user
-    NSDictionary* opts = @{static_cast<id>(kAXTrustedCheckOptionPrompt): @YES};
-    return AXIsProcessTrustedWithOptions(static_cast<CFDictionaryRef>(opts));
+#if __clang_major__ >= 9 && MAC_OS_X_VERSION_MIN_REQUIRED >= 1090
+    if (@available(macOS 10.15, *)) {
+        // Request accessibility permissions for Auto-Type type on behalf of the user
+        NSDictionary* opts = @{static_cast<id>(kAXTrustedCheckOptionPrompt): @YES};
+        return AXIsProcessTrustedWithOptions(static_cast<CFDictionaryRef>(opts));
+    }
+#else
+    return YES;
+#endif
 }
 
 //
--
Subject: [PATCH] Fix compilation when osx <= 10.7

* AppKitImpl.mm: The code uses @available syntax which is new in clang 9
This prevents compilation on older version of macOS that don't
use this version. For example on El Capitan.

* AppKitImpl.mm: CGDisplayStreamRef exists from 10.8 onwards only. It is used for a
  feature of KeePassXC that is only available from 10.15 onwards. So we don't need it
  when compiling on <= 10.7
error: unknown type name 'CGDisplayStreamRef'
--- src/gui/osutils/macutils/AppKitImpl.mm
+++ src/gui/osutils/macutils/AppKitImpl.mm
@@ -176,6 +178,7 @@ - (bool) enableAccessibility
 //
 - (bool) enableScreenRecording
 {
+#if __clang_major__ >= 9 && MAC_OS_X_VERSION_MIN_REQUIRED >= 1080
     if (@available(macOS 10.15, *)) {
         // Request screen recording permission on macOS 10.15+
         // This is necessary to get the current window title
@@ -193,6 +196,7 @@ - (bool) enableScreenRecording
             return NO;
         }
     }
+#endif
     return YES;
 }
 
--
Subject: [PATCH] fix compilation on Qt not having QOperatingSystemVersion::MacOSBigSur

The code uses 'QOperatingSystemVersion::MacOSBigSur' which doesn't exist
in all Qt versions (it has been backported to Qt 5.12.10+ & 5.15.1+ only).
On older macos systems like El Capitan the last supported
version of Qt is 5.11

This will fix compilation issue on such older systems and on systems
running with Qt not supporting QOperatingSystemVersion::MacOSBigSur

Compilation error was:
error: no member named 'MacOSBigSur' in 'QOperatingSystemVersion'
--- src/gui/styles/base/BaseStyle.cpp
+++ src/gui/styles/base/BaseStyle.cpp
@@ -53,8 +53,10 @@
 #include <qdrawutil.h>
 
 #ifdef Q_OS_MACOS
+#if QT_VERSION >= QT_VERSION_CHECK(5, 9, 0)
 #include <QOperatingSystemVersion>
 #endif
+#endif
 
 #include <cmath>
 
@@ -279,16 +281,22 @@ namespace Phantom
 #ifdef Q_OS_MACOS
             QColor tabBarBase(const QPalette& pal)
             {
+#if QT_VERSION >= QT_VERSION_CHECK(5, 12, 10) && QT_VERSION < QT_VERSION_CHECK(5, 13, 0)                               \
+    || QT_VERSION >= QT_VERSION_CHECK(5, 15, 1)
                 if (QOperatingSystemVersion::current() >= QOperatingSystemVersion::MacOSBigSur) {
                     return hack_isLightPalette(pal) ? QRgb(0xD4D4D4) : QRgb(0x2A2A2A);
                 }
+#endif
                 return hack_isLightPalette(pal) ? QRgb(0xDD1D1D1) : QRgb(0x252525);
             }
             QColor tabBarBaseInactive(const QPalette& pal)
             {
+#if QT_VERSION >= QT_VERSION_CHECK(5, 12, 10) && QT_VERSION < QT_VERSION_CHECK(5, 13, 0)                               \
+    || QT_VERSION >= QT_VERSION_CHECK(5, 15, 1)
                 if (QOperatingSystemVersion::current() >= QOperatingSystemVersion::MacOSBigSur) {
                     return hack_isLightPalette(pal) ? QRgb(0xF5F5F5) : QRgb(0x2D2D2D);
                 }
+#endif
                 return hack_isLightPalette(pal) ? QRgb(0xF4F4F4) : QRgb(0x282828);
             }
 #endif
--
Subject: [PATCH] Fix compilation when osx <= 10.7

* MacUtils.cpp: CoreGraphics exists from 10.8 onwards only, capslock detection feature
  would have to be implemented on OSX < 10.7

--- src/gui/osutils/macutils/MacUtils.cpp
+++ src/gui/osutils/macutils/MacUtils.cpp
@@ -24,8 +24,9 @@
 #include <QStandardPaths>
 #include <QTimer>
 
+#if MAC_OS_X_VERSION_MIN_REQUIRED >= 1080
 #include <CoreGraphics/CGEventSource.h>
-
+#endif
 
 QPointer<MacUtils> MacUtils::m_instance = nullptr;
 
@@ -136,7 +137,11 @@
 
 bool MacUtils::isCapslockEnabled()
 {
+#if MAC_OS_X_VERSION_MIN_REQUIRED >= 1080
     return (CGEventSourceFlagsState(kCGEventSourceStateHIDSystemState) & kCGEventFlagMaskAlphaShift) != 0;
+#else
+	return false;
+#endif
 }
 
 /**
